### PR TITLE
Introduce a new BinaryStreamSink that allows reading a stream of encoded bytes

### PR DESCRIPTION
**Pull request by jleibs**
### What
- Resolves: https://github.com/rerun-io/rerun/issues/6241

This effectively combines the guts of FileSink and MemorySink. Like the FileSink, this runs file-encoding on its own thread so that all we need to do when we're ready to read is move out the bytes.

The downside relative to memory_sink is the stream is a singular RRD. You have no way of accessing information about individual messages, draining from the backlog, etc.

This is designed to be as easy as possible to wire into a gradio output stream.

Example usage in a Gradio Component:
```python
@rr.thread_local_stream("rerun_example_live")
def start_stream_direct(state):

    state['streaming'] = True

    cap = cv2.VideoCapture(0)
    frame_nr = 0
    stream = rr.binary_stream()

    while state.get('streaming', False):
        ret, img = cap.read()

        frame_time_ms = cap.get(cv2.CAP_PROP_POS_MSEC)
        if frame_time_ms != 0:
            rr.set_time_nanos("frame_time", int(frame_time_ms * 1_000_000))

        rr.set_time_sequence("frame_nr", frame_nr)
        frame_nr += 1

        print("Processing frame", frame_nr)

        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

        # Log the original image
        rr.log("image/rgb", rr.Image(img).compress())

        # Convert to grayscale
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        rr.log("image/gray", rr.Image(gray).compress())

        # Run the canny edge detector
        canny = cv2.Canny(gray, 50, 200)
        rr.log("image/canny", rr.Image(canny).compress())

        yield stream.read()
```

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using examples from latest `main` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6242?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [rerun.io/viewer](https://rerun.io/viewer/pr/6242?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG
* [x] If applicable, add a new check to the [release checklist](https://github.com/rerun-io/rerun/blob/main/tests/python/release_checklist)!

- [PR Build Summary](https://build.rerun.io/pr/6242)
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

To run all checks from `main`, comment on the PR with `@rerun-bot full-check`.

### CODE REVIEW

File: crates/re_sdk/src/binary_stream_sink.rs [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    Send(LogMsg),
+    Flush(SyncSender<()>),
+}
+
+impl Command {
+    fn flush() -> (Self, Receiver<()>) {
+        let (tx, rx) = std::sync::mpsc::sync_channel(0); // oneshot
+        (Self::Flush(tx), rx)
+    }
+}
```
Comment (Wumpf) [inline crates/re_sdk/src/binary_stream_sink.rs:33]: nit: I found 3 implementations of this, two of them using crossbeam one using mpsc like you here. I like that you bring balance to the world ;)

File: crates/re_sdk/src/binary_stream_sink.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+impl BinaryStreamSink {
+    /// Create a pair of a new [`BinaryStreamSink`] and the associated [`BinaryStreamStorage`].
+    pub fn new(rec: RecordingStream) -> Result<(Self, BinaryStreamStorage), BinaryStreamSinkError> {
+        let storage = BinaryStreamStorage::new(rec);
+
+        // We always compress when writing to a stream
+        // TODO(jleibs): Make this configurable
```
Comment (Wumpf) [inline crates/re_sdk/src/binary_stream_sink.rs:131]: any harm in just exposing that as a flag on creation right away?

File: tests/python/binary_stream/main.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    for t in threads:
+        t.start()
+    for t in threads:
+        t.join()
+
+    while not results_queue.empty():
+        name, data = results_queue.get()
+
+        with open(f"output_{name}.rrd", "a+b") as f:
+            f.write(data)
```
Comment (Wumpf) [inline tests/python/binary_stream/main.py:59]: if you were to order the output of the queue and write a second rrd file with the same data in sync, then you could binary compare those two files, thus creating an automated test that we can put on ci.
This is too close to that to not do it!

**Review by Wumpf — CHANGES_REQUESTED** [review]
Neat utility, seems pretty straight forward

wondering if we should expose this new type of stream to C++ as well, but I reckon we don't have a usecase for it it, so better not to

Requesting changes because I'd love to see the test being automated - if this turns out to be too much work all good from my side

File: crates/re_sdk/src/binary_stream_sink.rs (hunk 1 above)
Comment (jleibs) [reply crates/re_sdk/src/binary_stream_sink.rs:33]: Good enough for FileSink, good enough for me :shrug:

File: crates/re_sdk/src/binary_stream_sink.rs (hunk 2 above)
Comment (jleibs) [reply crates/re_sdk/src/binary_stream_sink.rs:131]: Mostly that (1) it also applies to every other sink type and (2) I can't come up with a good reason that a user would ever set it to False.

I'd rather introduce them all together in a consistent way if and when  we have a compelling use-case.

File: tests/python/binary_stream/main.py (hunk 3 above)
Comment (jleibs) [reply tests/python/binary_stream/main.py:59]: Kind of annoying but I'll try to take a shot.

The rrd binary streams are never directly comparable due to randomness in log_time and tow tuid, which means I still need to write out to file and then shell out to rerun_compare.

If we had a python API for retrieving messages back from a stream I totally would have made it a proper unit-test from the start.

**Review by Wumpf — APPROVED** [review]
nice work! super cool to have this running on ci

**Pull request merged**